--- /home/nick/Projects/junk/safe/afl-1.83b/qemu_mode/build_qemu_support.sh	2015-04-29 20:34:59.000000000 -0700
+++ build_qemu_support.sh	2015-06-23 21:42:38.824340893 -0700
@@ -150,36 +150,70 @@
 cd ..
 ls -l ../afl-qemu-trace || exit 1
 
-echo "[+] Successfully created '../afl-qemu-trace'."
-
-echo "[*] Testing the build..."
-
-cd ..
-
-make >/dev/null || exit 1
-
-gcc test-instr.c -o test-instr || exit 1
+#echo "[+] Successfully created '../afl-qemu-trace'."
+#
+#echo "[*] Testing the build..."
+#
+#cd ..
+#
+#make >/dev/null || exit 1
+#
+#gcc test-instr.c -o test-instr || exit 1
+#
+#unset AFL_INST_RATIO
+#
+#echo 0 | ./afl-showmap -m none -Q -q -o .test-instr0 ./test-instr || exit 1
+#echo 1 | ./afl-showmap -m none -Q -q -o .test-instr1 ./test-instr || exit 1
+#
+#rm -f test-instr
+#
+#cmp -s .test-instr0 .test-instr1
+#DR="$?"
+#
+#rm -f .test-instr0 .test-instr1
+#
+#if [ "$DR" = "0" ]; then
+#
+#  echo "[-] Error: afl-qemu-trace instrumentation doesn't seem to work!"
+#  exit 1
+#
+#fi
+#
+#echo "[+] Instrumentation tests passed. "
+
+echo "[+] Building CGC qemu!"
+
+QEMU_DIR=qemu-cgc-afl
+rm -rf $QEMU_DIR
+echo "[*] Cloning our QEMU repository..."
+git clone git@git.seclab.cs.ucsb.edu:cgc/qemu.git $QEMU_DIR || exit 1
+echo "[*] Checking out our current afl pre-patched QEMU version..."
+git -C $QEMU_DIR checkout afl || exit 1
+git -C $QEMU_DIR pull || exit 1
+echo "[+] Checked out."
+
+echo "[*] Configuring QEMU..."
+
+cd $QEMU_DIR || exit 1
+
+./cgc_configure_opt
+
+echo "[+] Configuration complete."
+
+echo "[*] Attempting to build QEMU (fingers crossed!)..."
 
-unset AFL_INST_RATIO
+make -j7 || exit 1
 
-echo 0 | ./afl-showmap -m none -Q -q -o .test-instr0 ./test-instr || exit 1
-echo 1 | ./afl-showmap -m none -Q -q -o .test-instr1 ./test-instr || exit 1
+echo "[+] Build process successful!"
 
-rm -f test-instr
+echo "[*] Copying binary..."
 
-cmp -s .test-instr0 .test-instr1
-DR="$?"
+cp -f "i386-linux-user/qemu-i386" "../../afl-qemu-trace-cgc" || exit 1
 
-rm -f .test-instr0 .test-instr1
-
-if [ "$DR" = "0" ]; then
-
-  echo "[-] Error: afl-qemu-trace instrumentation doesn't seem to work!"
-  exit 1
-
-fi
+cd ..
+ls -l ../afl-qemu-cgc-trace || exit 1
 
-echo "[+] Instrumentation tests passed. "
+echo "[+] Successfully created '../afl-qemu-trace'."
 
 echo "[+] All set, you can now use the -Q mode in afl-fuzz!"
 
