--- /home/nick/Projects/junk/afl-1.83b/qemu_mode/patches/afl-qemu-cpu-inl.h	2015-06-10 23:33:30.000000000 -0700
+++ patches/afl-qemu-cpu-inl.h	2015-07-21 18:09:27.475532396 -0700
@@ -80,8 +80,8 @@
 
 /* Function declarations. */
 
-static void afl_setup(void);
-static void afl_forkserver(CPUArchState*);
+void afl_setup(void);
+void afl_forkserver(CPUArchState*);
 static inline void afl_maybe_log(abi_ulong);
 
 static void afl_wait_tsl(CPUArchState*, int);
@@ -107,7 +107,7 @@
 
 /* Set up SHM region and initialize other stuff. */
 
-static void afl_setup(void) {
+void afl_setup(void) {
 
   char *id_str = getenv(SHM_ENV_VAR),
        *inst_r = getenv("AFL_INST_RATIO");
@@ -154,7 +154,7 @@
 
 /* Fork server logic, invoked once we hit _start. */
 
-static void afl_forkserver(CPUArchState *env) {
+void afl_forkserver(CPUArchState *env) {
 
   static unsigned char tmp[4];
 
